colexec: add external hash joiner

yuzefovich opened the pull request:
**colexec: clean up hashing utility functions**

We have three hashing utility functions that previously had hashTable as
the receiver. However, this is not necessary since they do not rely on
anything in the hash table. This commit makes them standalone. The only
change needed to support that was introduction of cancelChecker as an
argument to `rehash`. This allowed removing the reference to "not fully
initialized hash table" from the hash router.

`bucketSize` variable has been renamed to `numBuckets` (since that's
what it actually means). This commit also adds support for `numBuckets`
values that are not power of 2 to `finalizeHash` function.

Release note: None

**colexec: miscellaneous cleanups and additions**

This commit removes `Eq` from the name of hash join operator. It also
adds several interface implementation assertions and removes a stale
entry from .gitignore file. A new function is introduced into the hash
table to help with the reuse of some allocations in the follow-up PRs.

Next, this commit removes `hashJoinProber` entirely and moves it's
contents into `hashJoiner` into an embedded struct. The reason for this
change is that some fields were duplicated in both structs and there
would be no clear ownership of who should be resetting the hash table.

`reset` method has been added to the hash table and hash join operator.
`bufferingInMemoryOperator`'s method `ExportBuffered` has been modified
to take an argument to support the case of a buffering operator with
several inputs (hash join, for example). The argument will be used to
distinguish which input to "export the buffered tuples from".

**Important change:** the probing code has been changed to not put
`Vec.Copy` calls through the allocator (i.e. when we're probing and
populating the output batch, `Copy` methods are not accounted for). This
is done for several reasons:
1. we have already built the hash table, so it would be a shame to fall
back to disk-backed algorithm at this point
2. actually falling back in such scenario is non-trivial because we
might have emitted partial matches from the current probing batch
3. the deviation from the memory limit should not be significant because
the output batch has already been allocated, so the only concern is
about variable-width types - if our estimation significantly
underestimates the memory footprint of variable-width type vectors, we
will be actually using more memory than what we accounted for.

At this point I think it is an acceptable compromise, but it should be
addressed before the release.

Release note: None

**colexec: add two input disk spiller**

This commit adds a disk spilling operator for an in-memory operator that
operates on two inputs. Most of the logic is shared with
oneInputDiskSpiller, only execinfra.OpNode methods are slightly
different.

Release note: None

**colexec: factor out some hash utilities**

This commit extracts out the logic of "tuple distribution" out of hash
router into a helper struct which will be reused by the external hash
join. It also moves 3 hashing utility functions out of `hashtable` file
into `hash_utils`.

Additionally, now hash router stores `hashCols` as `[]uint32` since
that's what is used in the protos.

Release note: None

**colexec: add external hash joiner**

This commit adds external hash joiner that implements Grace hash join
algorithm. It is done in two phases:
1. we read the batches from both inputs and partition them according to
the vectorized hash function initialized with 2's (unlike default value
of 1)
2. once we have exhausted both inputs, we proceed to join phase in which
we reuse the in-memory hash joiner to actually perform a join operation
on the tuples from the left and right sides that hashed to the same
partition. Here, the in-memory hash joiner uses the default value of
1 when initializing the hash values, so essentially it uses a different
hash function.

Current limitations:
- we use fixed number of buckets (4096) in the partitioning phase
- we do not support recursive partitioning, i.e. if we have a partition
that will not fit into memory, we will just bail on the query.

Fixes: #43790.

Release note: None

**colexec: fix repeatable batch source**

RepeatableBatchSource is a utility operator that returns the batch it's
initialized with forever. Previously it would only store the length and
the selection vector of the input batch separately, but this is not
sufficient - downstream operators are allowed to modify the batch which
is output by the repeatable batch source in any way, including
overwriting the actual data. Now this is fixed by storing the contents
of the input batch and populating a separate output batch on every call
to Next.

Release note: None

yuzefovich committed:
colexec: clean up hashing utility functions

We have three hashing utility functions that previously had hashTable as
the receiver. However, this is not necessary since they do not rely on
anything in the hash table. This commit makes them standalone. The only
change needed to support that was introduction of cancelChecker as an
argument to `rehash`. This allowed removing the reference to "not fully
initialized hash table" from the hash router.

`bucketSize` variable has been renamed to `numBuckets` (since that's
what it actually means). This commit also adds support for `numBuckets`
values that are not power of 2 to `finalizeHash` function.

Release note: None

yuzefovich committed:
colexec: miscellaneous cleanups and additions

This commit removes `Eq` from the name of hash join operator. It also
adds several interface implementation assertions and removes a stale
entry from .gitignore file. A new function is introduced into the hash
table to help with the reuse of some allocations in the follow-up PRs.

Next, this commit removes `hashJoinProber` entirely and moves it's
contents into `hashJoiner` into an embedded struct. The reason for this
change is that some fields were duplicated in both structs and there
would be no clear ownership of who should be resetting the hash table.

`reset` method has been added to the hash table and hash join operator.
`bufferingInMemoryOperator`'s method `ExportBuffered` has been modified
to take an argument to support the case of a buffering operator with
several inputs (hash join, for example). The argument will be used to
distinguish which input to "export the buffered tuples from".

**Important change:** the probing code has been changed to not put
`Vec.Copy` calls through the allocator (i.e. when we're probing and
populating the output batch, `Copy` methods are not accounted for). This
is done for several reasons:
1. we have already built the hash table, so it would be a shame to fall
back to disk-backed algorithm at this point
2. actually falling back in such scenario is non-trivial because we
might have emitted partial matches from the current probing batch
3. the deviation from the memory limit should not be significant because
the output batch has already been allocated, so the only concern is
about variable-width types - if our estimation significantly
underestimates the memory footprint of variable-width type vectors, we
will be actually using more memory than what we accounted for.

At this point I think it is an acceptable compromise, but we might want
to address it before the release.

Release note: None

yuzefovich committed:
colexec: add two input disk spiller

This commit refactors oneInputDiskSpiller into diskSpillerBase that can
support variable number of inputs.

Release note: None

yuzefovich committed:
colexec: factor out some hash utilities

This commit extracts out the logic of "tuple distribution" out of hash
router into a helper struct which will be reused by the external hash
join. It also moves 3 hashing utility functions out of `hashtable` file
into `hash_utils`.

Additionally, now hash router stores `hashCols` as `[]uint32` since
that's what is used in the protos.

Release note: None

yuzefovich committed:
colexec: add external hash joiner

This commit adds external hash joiner that implements Grace hash join
algorithm. It is done in two phases:
1. we read the batches from both inputs and partition them according to
the vectorized hash function initialized with 2's (unlike default value
of 1)
2. once we have exhausted both inputs, we proceed to join phase in which
we reuse the in-memory hash joiner to actually perform a join operation
on the tuples from the left and right sides that hashed to the same
partition. Here, the in-memory hash joiner uses the default value of
1 when initializing the hash values, so essentially it uses a different
hash function.

Current limitations:
- we use fixed number of buckets (4096) in the partitioning phase
- we do not support recursive partitioning, i.e. if we have a partition
that will not fit into memory, we will just bail on the query.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44523)
<!-- Reviewable:end -->

asubiotto commented:
Great to see this! Aiming to fully review this this afternoon.

asubiotto reviewed (changes requested):
Haven't managed to review the last commit yet

> Reviewed 8 of 8 files at r1, 9 of 9 files at r2, 1 of 1 files at r3, 12 of 12 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-:-LzsI8_KDf5bhZ0YzGkG:by1z0h0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/disk_spiller.go, line 105 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzs57ZtExSgN0wH_x_p:-Lzs57ZtExSgN0wH_x_q:b-s9fcfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e03c9ff8b9cfa05a52db28b74686042cc477238d/pkg/sql/colexec/disk_spiller.go#L105)):*
> ```Go
> ) Operator {
> 	diskBackedOpInput := newBufferExportingOperator(inMemoryOp, input)
> 	base := diskSpillerBase{
> ```

nit: I usually find it easier to read if the construction of `diskSpillerBase` is embedded in `oneInputDiskSpiller`
___
*[pkg/sql/colexec/disk_spiller.go, line 148 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzs7QusECOuhkZsYB6Q:-Lzs7QusECOuhkZsYB6R:b-sjg8hp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e03c9ff8b9cfa05a52db28b74686042cc477238d/pkg/sql/colexec/disk_spiller.go#L148)):*
> ```Go
> //   will simply propagate the batch to the output. If that fails with an OOM
> //   error, the disk spiller will then initialize the right side chain and will
> //   proceed to emit from there
> ```

nit: since you're in this file anyways, it seems like this comment is missing periods (as well as the mirrored one above)
___
*[pkg/sql/colexec/disk_spiller.go, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzs99qV1JSKJakxwUMv:-Lzs99qV1JSKJakxwUMw:b-yz3whp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e03c9ff8b9cfa05a52db28b74686042cc477238d/pkg/sql/colexec/disk_spiller.go#L174)):*
> ```Go
> // - spillingCallbackFn will be called when the spilling from in-memory to disk
> //   backed operator occurs. It should only be set in tests.
> func newTwoInputDiskSpiller(
> ```

Instead of creating a new struct, could you just use a `[]Operator` that you assert is either of length 1 or two? It seems like we can just add some extra logic to `Child` to behave properly (i.e. return something based on the length of `inputs []Operator`
___
*[pkg/sql/colexec/execplan.go, line 661 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsH3HA6C6zjz8mI-cU:-LzsH3HA6C6zjz8mI-cV:b-u60ze1) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ceb48c6cd4d6ace30bffae73a984c7b74b4bfe/pkg/sql/colexec/execplan.go#L661)):*
> ```Go
> 					rightEqColsAreKey,
> 				)
> 				if err == nil {
> ```

I think it's easier to read to put the `if err != nil` case here.
___
*[pkg/sql/colexec/hash_utils.go, line 26 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsAkEo3NxHKm_LgS6-:-LzsAkEo3NxHKm_LgS60:b4dub4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bff8bd6b00c22b076edd21d5b3b0ef910f02f77/pkg/sql/colexec/hash_utils.go#L26)):*
> ```Go
> //
> // We currently use the same hash functions used by go's maps.
> // TODO(asubiotto): Once https://go-review.googlesource.com/c/go/+/155118/ is
> ```

Looks like they added the `maphash` package in go1.14. This gets us partially there by allowing us not to define the hash function themselves, but we'll still need to cast to bytes and use noescape I believe. I filed #44557 
___
*[pkg/sql/colexec/hash_utils.go, line 29 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsCcPl0RwKdaWBRzzu:-LzsCcPl0RwKdaWBRzzv:b-o0iyzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bff8bd6b00c22b076edd21d5b3b0ef910f02f77/pkg/sql/colexec/hash_utils.go#L29)):*
> ```Go
> // in, we should use the public API.
> 
> const defaultInitHashValue = 1
> ```

Please add a comment.
___
*[pkg/sql/colexec/hash_utils.go, line 67 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsCMsw70XZYD6BDcF5:-LzsCMsw70XZYD6BDcF6:b-sgyc9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bff8bd6b00c22b076edd21d5b3b0ef910f02f77/pkg/sql/colexec/hash_utils.go#L67)):*
> ```Go
> // according to the corresponding hashes. The "distribution" occurs by
> // populating selection vectors which the caller needs to use accordingly.
> type tupleHashDistributor struct {
> ```

This is a nice idea. @azhng you might want to reuse this for the hash aggregator.
___
*[pkg/sql/colexec/hashjoiner.go, line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzrz6Nt0czV9gw8qshR:-Lzrz6Nt0czV9gw8qshS:bjxoejc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ff04ee49ec070eeb8bba2f17f34d0f737a04de/pkg/sql/colexec/hashjoiner.go#L323)):*
> ```Go
> 		// to hit the limit here, it would have been very hard to fall back to disk
> 		// backed hash joiner because we might have already emitted partial output.
> 		// TODO(yuzefovich): I think this behavior is acceptable - we allocated
> ```

I think this behavior is acceptable too.
___
*[pkg/sql/colexec/hashjoiner.go, line 576 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzs0tYM09HMLgXuIq03:-Lzs0tYM09HMLgXuIq04:b-a7lprm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ff04ee49ec070eeb8bba2f17f34d0f737a04de/pkg/sql/colexec/hashjoiner.go#L576)):*
> ```Go
> 		}
> 	}
> 	// TODO(yuzefovich): hash table does *not* release the memory under
> ```

Could you describe this more in detail? I'm not sure why we wouldn't want release the memory.
___
*[pkg/sql/colexec/routers.go, line 427 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsDgwpBjvlcIUUFU_t:-LzsDgwpBjvlcIUUFU_u:b-9mo1yl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bff8bd6b00c22b076edd21d5b3b0ef910f02f77/pkg/sql/colexec/routers.go#L427)):*
> ```Go
> 	}
> 
> 	selections := r.tupleDistributor.distribute(ctx, b, r.types, r.hashCols)
> ```

Since a `tupleDistributor` sort of assumes a schema when instantiated, I think it might be nicer to just make `r.types` and `r.hashCol` part of the `tupleDistributor` constructor and just pass in `ctx, b` here


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Added a first commit to - hopefully - make the build green. Also moved the test for "hashing function family" into the second to last commit from the last one since it belongs there.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-:-LztVOdo-Oyu3MhHFUlW:btyr0j9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/disk_spiller.go, line 105 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzs57ZtExSgN0wH_x_p:-LztNmGv3HQ3B17hjGVZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e03c9ff8b9cfa05a52db28b74686042cc477238d/pkg/sql/colexec/disk_spiller.go#L105)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I usually find it easier to read if the construction of `diskSpillerBase` is embedded in `oneInputDiskSpiller`
</blockquote></details>

Done.
___
*[pkg/sql/colexec/disk_spiller.go, line 148 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzs7QusECOuhkZsYB6Q:-LztOUU401AzLKj2xys5:b-5p2su7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e03c9ff8b9cfa05a52db28b74686042cc477238d/pkg/sql/colexec/disk_spiller.go#L148)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: since you're in this file anyways, it seems like this comment is missing periods (as well as the mirrored one above)
</blockquote></details>

Done.

I didn't put the periods at first because I considered this whole paragraph as a single sentence, but I don't have strong preference here.
___
*[pkg/sql/colexec/disk_spiller.go, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzs99qV1JSKJakxwUMv:-LztOgSZ88i5vlHVoTPB:b1a98vs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e03c9ff8b9cfa05a52db28b74686042cc477238d/pkg/sql/colexec/disk_spiller.go#L174)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Instead of creating a new struct, could you just use a `[]Operator` that you assert is either of length 1 or two? It seems like we can just add some extra logic to `Child` to behave properly (i.e. return something based on the length of `inputs []Operator`
</blockquote></details>

Good point, refactored.

I decided to keep the diagram since it can be useful to understand the behavior.
___
*[pkg/sql/colexec/execplan.go, line 661 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsH3HA6C6zjz8mI-cU:-LztQXvIE21uytiRkdoO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ceb48c6cd4d6ace30bffae73a984c7b74b4bfe/pkg/sql/colexec/execplan.go#L661)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think it's easier to read to put the `if err != nil` case here.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/hash_utils.go, line 26 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsAkEo3NxHKm_LgS6-:-LztJFIU3mTgNswlyctv:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bff8bd6b00c22b076edd21d5b3b0ef910f02f77/pkg/sql/colexec/hash_utils.go#L26)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Looks like they added the `maphash` package in go1.14. This gets us partially there by allowing us not to define the hash function themselves, but we'll still need to cast to bytes and use noescape I believe. I filed #44557 
</blockquote></details>

Ack.
___
*[pkg/sql/colexec/hash_utils.go, line 29 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsCcPl0RwKdaWBRzzu:-LztPj8I2vkXYA2H2rU4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bff8bd6b00c22b076edd21d5b3b0ef910f02f77/pkg/sql/colexec/hash_utils.go#L29)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Please add a comment.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/hashjoiner.go, line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzrz6Nt0czV9gw8qshR:-LztJju4BNAEV9u96PhW:bc8r00n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ff04ee49ec070eeb8bba2f17f34d0f737a04de/pkg/sql/colexec/hashjoiner.go#L323)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think this behavior is acceptable too.
</blockquote></details>

Cool, removed the todo.
___
*[pkg/sql/colexec/hashjoiner.go, line 576 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-Lzs0tYM09HMLgXuIq03:-LztKgjl1NEnmEeTAzIw:blff9u4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3ff04ee49ec070eeb8bba2f17f34d0f737a04de/pkg/sql/colexec/hashjoiner.go#L576)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Could you describe this more in detail? I'm not sure why we wouldn't want release the memory.
</blockquote></details>

~The idea here is that if we're resetting the in-memory hash joiner, then we intend to use it multiple times. On the first partition hash joiner will be allocating memory to hold the whole right side of the partition (as `hj.ht.vals` which is a `bufferedBatch`). On the second and following partitions we would prefer to reuse that already allocated memory for performance reasons, so we're holding on to the memory in `reset`. If we were to let go off it, we would have to allocate the memory for hash table on every partition.~

~At the same time, currently, if we do not clear the allocator, when we will start buffering up tuples from the second partition, the allocator will be thinking that all memory under the buffered batch has been released because we're setting `bufferedBatch.length=0`.~

Actually, I was wrong - our `PerformOperation` is smart enough to look at the capacities (and not lengths) of the vectors, so we don't need to update the memory account of the allocator at all :)
___
*[pkg/sql/colexec/routers.go, line 427 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-LzsDgwpBjvlcIUUFU_t:-LztIqjK9IQ5cLO260oF:bkman4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bff8bd6b00c22b076edd21d5b3b0ef910f02f77/pkg/sql/colexec/routers.go#L427)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Since a `tupleDistributor` sort of assumes a schema when instantiated, I think it might be nicer to just make `r.types` and `r.hashCol` part of the `tupleDistributor` constructor and just pass in `ctx, b` here
</blockquote></details>

That's what I did initially. But then I realized that in case of hash joiner, two inputs can have different schemas, so in order to reuse the same `tupleDistributor` I put `types` and `hashCols` as arguments to give more flexibility.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: fix repeatable batch source

RepeatableBatchSource is a utility operator that returns the batch it's
initialized with forever. Previously it would only store the length and
the selection vector of the input batch separately, but this is not
sufficient - downstream operators are allowed to modify the batch which
is output by the repeatable batch source in any way, including
overwriting the actual data. Now this is fixed by storing the contents
of the input batch and populating a separate output batch on every call
to Next.

Release note: None

yuzefovich commented:
Added another commit that fixes the benchmarks. Here are the numbers:
```
BenchmarkExternalHashJoiner/nulls=false/fullOuter=false/batches=4/spilled=false-16         	    2838	    376919 ns/op	 486.30 MB/s
BenchmarkExternalHashJoiner/nulls=false/fullOuter=false/batches=4/spilled=true-16          	    2928	    374470 ns/op	 489.48 MB/s
BenchmarkExternalHashJoiner/nulls=false/fullOuter=false/batches=64/spilled=false-16        	      42	  29317589 ns/op	 100.03 MB/s
BenchmarkExternalHashJoiner/nulls=false/fullOuter=false/batches=64/spilled=true-16         	      40	  28763293 ns/op	 101.96 MB/s
BenchmarkExternalHashJoiner/nulls=false/fullOuter=true/batches=4/spilled=false-16          	    2842	    428705 ns/op	 427.56 MB/s
BenchmarkExternalHashJoiner/nulls=false/fullOuter=true/batches=4/spilled=true-16           	    2749	    429455 ns/op	 426.81 MB/s
BenchmarkExternalHashJoiner/nulls=false/fullOuter=true/batches=64/spilled=false-16         	      31	  38608804 ns/op	  75.96 MB/s
BenchmarkExternalHashJoiner/nulls=false/fullOuter=true/batches=64/spilled=true-16          	      30	  38000272 ns/op	  77.18 MB/s
BenchmarkExternalHashJoiner/nulls=true/fullOuter=false/batches=4/spilled=false-16          	    2188	    566526 ns/op	 323.54 MB/s
BenchmarkExternalHashJoiner/nulls=true/fullOuter=false/batches=4/spilled=true-16           	    2010	    546183 ns/op	 335.59 MB/s
BenchmarkExternalHashJoiner/nulls=true/fullOuter=false/batches=64/spilled=false-16         	      30	  39841123 ns/op	  73.61 MB/s
BenchmarkExternalHashJoiner/nulls=true/fullOuter=false/batches=64/spilled=true-16          	      30	  39759820 ns/op	  73.76 MB/s
BenchmarkExternalHashJoiner/nulls=true/fullOuter=true/batches=4/spilled=false-16           	    1935	    577482 ns/op	 317.41 MB/s
BenchmarkExternalHashJoiner/nulls=true/fullOuter=true/batches=4/spilled=true-16            	    2109	    582520 ns/op	 314.66 MB/s
BenchmarkExternalHashJoiner/nulls=true/fullOuter=true/batches=64/spilled=false-16          	      22	  49266073 ns/op	  59.53 MB/s
BenchmarkExternalHashJoiner/nulls=true/fullOuter=true/batches=64/spilled=true-16           	      24	  49903841 ns/op	  58.77 MB/s
```
Note that comparing them directly against `BenchmarkHashJoiner` is not fair because that benchmark outputs *only* `nBatches` whereas `BenchmarkExternalHashJoiner` outputs all batches. That is also the reason why on `batches=64` the performance seems to be worse - the join simply outputs a lot more data.

asubiotto reviewed (changes requested):
> Reviewed 4 of 37 files at r6, 4 of 8 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-Ar0vK8gtV5wsXF6aN:-M-Ar0vK8gtV5wsXF6aO:b-akzzhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner.go#L50)):*
> ```Go
> // TODO(yuzefovich): make this tunable. Ideally, we would calculate this number
> // based on the cardinality of inputs.
> const externalHJNumPartitions = 1 << 12
> ```

nit: I find it easier to read when the first digit is the same (e.g. in this case `4 << 10` makes me think of 4KiB instantly), but maybe that's just me so do what you prefer. In anyc case, it would probably be good to just have a `/* 4096 */` comment.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 165 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-At7vB3Ps-J7mU68lg:-M-At7vB3Ps-J7mU68lh:b-mmf3gb) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner.go#L165)):*
> ```Go
> 		if len(sel) > 0 {
> 			batch.SetSelection(true)
> 			copy(batch.Selection(), sel)
> ```

I think this is going to be a problem. The serialization code always assumes batches without selection vectors (because arrow doesn't have these), so this won't work. I think we'll have to copy to a new batch without a selection vector. It should be fine to reuse a batch for this purpose (for all partitions) and the copy itself shouldn't have much overhead wrt writing to disk.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 200 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-AvYXVBe9QcN3ToIwN:-M-AvYXVBe9QcN3ToIwO:b-e6b7ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner.go#L200)):*
> ```Go
> 				// All partitions have been processed, so we clean up the disk
> 				// infrastructure and transition to finished state.
> 				if err := hj.leftPartitioner.Close(); err != nil {
> ```

I've been wondering about this. How is this infrastructure cleaned up in other cases (e.g. the flow starts draining mid-flow)?
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 49 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-AvzauBky44voDnjOn:-M-Avzav0C0bnKLRzqGU:bysrgp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner_test.go#L49)):*
> ```Go
> 	// 0 - the default 64MiB value is used, so we exercise that the disk spilling
> 	//     machinery doesn't affect the correctness.
> 	// 1 - this will force the in-memory hash join operator to hit the memory
> ```

One test case i added to the hash router which might be good to add here (and maybe the sorter) is randomize a limit between 1 and something very large. We just want to assert that the hash joiner behaves
___
*[pkg/sql/colexec/external_sort.go, line 341 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-Awdnu7gi7cg43jnIB:-M-Awdnu7gi7cg43jnIC:b-pom2ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_sort.go#L341)):*
> ```Go
> 	if partition == nil {
> 		// TODO(yuzefovich): clarify the interface in this case.
> 		//return errors.Newf("partition %d not found (len(partitions) = %d)", partitionIdx, len(d.partitions))
> ```

nit: remove comment?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-Ar0vK8gtV5wsXF6aN:-M-BZAXGD8Qh5QMZ03kJ:bgbnler) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner.go#L50)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I find it easier to read when the first digit is the same (e.g. in this case `4 << 10` makes me think of 4KiB instantly), but maybe that's just me so do what you prefer. In anyc case, it would probably be good to just have a `/* 4096 */` comment.
</blockquote></details>

I agree, it's better, done.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 165 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-At7vB3Ps-J7mU68lg:-M-BamoS6tDvsOWhVgCP:b87yafc) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner.go#L165)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think this is going to be a problem. The serialization code always assumes batches without selection vectors (because arrow doesn't have these), so this won't work. I think we'll have to copy to a new batch without a selection vector. It should be fine to reuse a batch for this purpose (for all partitions) and the copy itself shouldn't have much overhead wrt writing to disk.
</blockquote></details>

Good point, fixed.

This change actually fixes an issue with benchmarks I was seeing, yet I think the fix of `RepeatableBatchSource` is beneficial to have.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 200 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-AvYXVBe9QcN3ToIwN:-M-BbNCjCm4aGnCOfYFB:b-u6cf65) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner.go#L200)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I've been wondering about this. How is this infrastructure cleaned up in other cases (e.g. the flow starts draining mid-flow)?
</blockquote></details>

That is a good point that we haven't discussed yet. I think we could talk about it tomorrow, I don't know the answer at the moment, but introducing an infrastructure for that cleanup should be a part of a separate PR. Opened up an issue for that.
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 49 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-AvzauBky44voDnjOn:-M-Bdy165Oi3M1iTKddt:b5s4qyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner_test.go#L49)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

One test case i added to the hash router which might be good to add here (and maybe the sorter) is randomize a limit between 1 and something very large. We just want to assert that the hash joiner behaves
</blockquote></details>

I don't think it would give any more coverage in the unit tests - behavior with random memory limit will be a combination of memory limits 0 and 1(if random value is less than `mon.DefaultPoolAllocation`, then only a single batch is buffered up by the in-memory hash joiner - `1` case, if it is greater, then the spilling will not occur - `0` case).
___
*[pkg/sql/colexec/external_sort.go, line 341 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-Awdnu7gi7cg43jnIB:-M-BeTzd5jEytHuInvtu:b-tv0yy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_sort.go#L341)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: remove comment?
</blockquote></details>

No, we actually need to think through this case.

The question is: what should we do when trying to dequeue from a partition to which we have not enqueued anything? I think it should be ok to return zero-length batch.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> Reviewed 3 of 8 files at r7, 34 of 34 files at r12, 9 of 9 files at r13, 1 of 1 files at r14, 13 of 13 files at r15, 10 of 10 files at r16, 23 of 23 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/disk_spiller.go, line 75 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HPbeJ3MF3awjXw8O_:-M-HPbeJ3MF3awjXw8Oa:bqh3crv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bae34b8c48b05e44cf14fb731a3066a9ad05adfa/pkg/sql/colexec/disk_spiller.go#L75)):*
> ```Go
> //   former will first export all the buffered tuples from inMemoryOp and then
> //   will proceed on emitting from input.
> //
> ```

nit: I'd expect this to just be an empty line (not a comment) to separate the constructor comment
___
*[pkg/sql/colexec/execplan.go, line 567 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HR-9_C9q3Z2wYMB9a:-M-HR-9_C9q3Z2wYMB9b:b-lgwaiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L567)):*
> ```Go
> 				if useStreamingMemAccountForBuffering {
> 					// We will not be creating a disk-backed hash joiner because we're
> 					// running a test that implicitly asked for only in-memory hash
> ```

Is this an implicit ask for in-memory hash joiner? Does this happen with the sorter too?
___
*[pkg/sql/colexec/execplan.go, line 576 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HQYcr02imHTtOdxI4:-M-HQYcr02imHTtOdxI5:b7cvn84) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L576)):*
> ```Go
> 						func(inputOne, inputTwo Operator) Operator {
> 							monitorNamePrefix := "external-hash-joiner-"
> 							// We unset the testing knob for memory limit because we external
> ```

I'm not sure I understand this comment. Could you try rephrasing it?
___
*[pkg/sql/colexec/execplan.go, line 583 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HR4shBSvG_4j5A3cz:-M-HR4shBSvG_4j5A3d-:b-442d2) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L583)):*
> ```Go
> 							// instantiated the "explicitly" in-memory hash join operator
> 							// above.
> 							flowCtx.Cfg.TestingKnobs.MemoryLimitBytes = 0
> ```

Should we reset it afterwards?
___
*[pkg/sql/colexec/external_hash_joiner.go, line 200 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-AvYXVBe9QcN3ToIwN:-M-HRHU_EZJSBrDoyY-4:b5n17kv) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner.go#L200)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

That is a good point that we haven't discussed yet. I think we could talk about it tomorrow, I don't know the answer at the moment, but introducing an infrastructure for that cleanup should be a part of a separate PR. Opened up an issue for that.
</blockquote></details>

Discussed offline to register `Closer`s with the flow that will be called during `Cleanup`.
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 49 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-AvzauBky44voDnjOn:-M-HOXlwEC5V0pf7aisy:brds1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner_test.go#L49)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I don't think it would give any more coverage in the unit tests - behavior with random memory limit will be a combination of memory limits 0 and 1(if random value is less than `mon.DefaultPoolAllocation`, then only a single batch is buffered up by the in-memory hash joiner - `1` case, if it is greater, then the spilling will not occur - `0` case).
</blockquote></details>

You might test a case where several batches are buffered before spilling to disk. The hash router has several heuristics that it uses to spill to disk, so maybe it's more useful in that case. Up to you.
___
*[pkg/sql/colexec/external_sort.go, line 341 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-Awdnu7gi7cg43jnIB:-M-HOrw1750bM-HhEJ9Z:btd51a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_sort.go#L341)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

No, we actually need to think through this case.

The question is: what should we do when trying to dequeue from a partition to which we have not enqueued anything? I think it should be ok to return zero-length batch.
</blockquote></details>

I agree.
___
*[pkg/sql/colexec/hash_utils.go, line 46 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HQ0qw4QHa_YzOq0h0:-M-HQ0qw4QHa_YzOq0h1:bfpx9sr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d024085207b4541c0bff8f02915e919562374cd5/pkg/sql/colexec/hash_utils.go#L46)):*
> ```Go
> // initHash initializes the hash value of each key to its initial state for
> // rehashing purposes.
> // NOTE: initValue *must* be non-zero. Different initValues
> ```

nit: unfinished comment?
___
*[pkg/sql/colexec/hash_utils_test.go, line 27 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HQ5QU2P5j7wYYG18w:-M-HQ5QU2P5j7wYYG18x:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d024085207b4541c0bff8f02915e919562374cd5/pkg/sql/colexec/hash_utils_test.go#L27)):*
> ```Go
> // defines a function family and that changing the initial hash value is
> // sufficient to get a "different" hash function.
> func TestHashFunctionFamily(t *testing.T) {
> ```

:+1: 
___
*[pkg/sql/colexec/testutils.go, line 74 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HS-RK3Sp-V5xdglo_:-M-HS-RK3Sp-V5xdgloa:bjealke) ([raw file](https://github.com/cockroachdb/cockroach/blob/0759e56c310bac8f7e110ba05e3806a0775d4f75/pkg/sql/colexec/testutils.go#L74)):*
> ```Go
> // input batch forever. Note that it stores the contents of the input batch and
> // copies them into a separate output batch. The output batch is allowed to be
> // modified whereas the input batch is *not*.
> ```

Is there a way we can check for modifications and only copy if so? I'm just slightly worried this will skew benchmarks.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/disk_spiller.go, line 75 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HPbeJ3MF3awjXw8O_:-M-I1RNcAhtjzEhlGgw2:bnxmh0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/bae34b8c48b05e44cf14fb731a3066a9ad05adfa/pkg/sql/colexec/disk_spiller.go#L75)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I'd expect this to just be an empty line (not a comment) to separate the constructor comment
</blockquote></details>

Done. I though it was weird to have a comment describing a struct that doesn't exist :)
___
*[pkg/sql/colexec/execplan.go, line 567 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HR-9_C9q3Z2wYMB9a:-M-I26De-JGk_ZkP9VED:b-j7xh7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L567)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is this an implicit ask for in-memory hash joiner? Does this happen with the sorter too?
</blockquote></details>

Yes, it is an implicit ask specifically for in-memory hash joiner to be used in unit tests of `hashJoiner` because we're calling `NewColOperator` to create it.

With the sorter the story is different at the moment - the unit test call `NewSorter` explicitly, so there is no question whether to add spilling.

I agree that this `useStreamingMemAccountForBuffering` flag becomes overloaded. Added another testing knob for this purpose.
___
*[pkg/sql/colexec/execplan.go, line 576 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HQYcr02imHTtOdxI4:-M-I6tApBkEw-kzzYa-U:b-1n4hqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L576)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm not sure I understand this comment. Could you try rephrasing it?
</blockquote></details>

I tried, not sure how to phrase it well. Basically the issue is that external hash joiner *must* have some memory to work with, and if we keep the testing knob's memory limit unchanged (which can be 1), then the external hash joiner can either fail on the construction or later internal in-memory hash join operator can hit OOM which would force to do the recursive partitioning which we don't have yet.
___
*[pkg/sql/colexec/execplan.go, line 583 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HR4shBSvG_4j5A3cz:-M-I7KwOEMJqdUE-SX7q:b-odvs3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L583)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should we reset it afterwards?
</blockquote></details>

I don't think it currently matters, but yes, done.
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 49 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-AvzauBky44voDnjOn:-M-I7WJT99fi0WVvq5aH:b-r205gd) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_hash_joiner_test.go#L49)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

You might test a case where several batches are buffered before spilling to disk. The hash router has several heuristics that it uses to spill to disk, so maybe it's more useful in that case. Up to you.
</blockquote></details>

I think I tried different memory limits with external sorter trying to spill after buffering several batches, but my attempt wasn't very successful. We can look into it later.
___
*[pkg/sql/colexec/external_sort.go, line 341 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-Awdnu7gi7cg43jnIB:-M-I8Mk-3LnnWemTRAGH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/723c9feb021e2dda5b2b8ad7aa61e0d3339d6a1b/pkg/sql/colexec/external_sort.go#L341)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I agree.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/hash_utils.go, line 46 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HQ0qw4QHa_YzOq0h0:-M-I8_dmD6f-KJWAmaJa:bd954vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d024085207b4541c0bff8f02915e919562374cd5/pkg/sql/colexec/hash_utils.go#L46)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: unfinished comment?
</blockquote></details>

A leftover - I put the comment I wanted to add here above (to highlight that changing init value changes the hash function). Fixed.
___
*[pkg/sql/colexec/testutils.go, line 74 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HS-RK3Sp-V5xdglo_:-M-IA0Fs6x_q7Rif6toM:bydh2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/0759e56c310bac8f7e110ba05e3806a0775d4f75/pkg/sql/colexec/testutils.go#L74)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is there a way we can check for modifications and only copy if so? I'm just slightly worried this will skew benchmarks.
</blockquote></details>

Hm, comparing whole vectors might be as expensive as copying the data directly. We could compute something like a checksum over all of the values in all vectors.

Another idea is we could add a flag to `RepeatableBatchSource` to have old behavior (something like `unsafe`), and the benchmarks that know for sure that the data in the output batch is not modified (maybe appended, or sel or length is changed, but not the values in the vectors) could set that flag.

Anyway, I'd rather have this go in as-is and introduce those things later if necessary.

I agree that the numbers in the benchmarks will be skewed, but to me that doesn't seem like a big concern to be honest.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
minus some minor nits <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 42 of 42 files at r18, 13 of 13 files at r19, 10 of 10 files at r20, 23 of 23 files at r21.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-:-M-Ks9TM3ddU1I4fusSk:b-3j7wmo)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 576 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HQYcr02imHTtOdxI4:-M-KrVdu3Fi0IQgEJrgz:bo1v4fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L576)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I tried, not sure how to phrase it well. Basically the issue is that external hash joiner *must* have some memory to work with, and if we keep the testing knob's memory limit unchanged (which can be 1), then the external hash joiner can either fail on the construction or later internal in-memory hash join operator can hit OOM which would force to do the recursive partitioning which we don't have yet.
</blockquote></details>

Got it, might be useful to mention the recursive partitioning.
___
*[pkg/sql/colexec/execplan.go, line 583 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HR4shBSvG_4j5A3cz:-M-Krq886KcaxHSXk8qf:b1cugzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L583)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I don't think it currently matters, but yes, done.
</blockquote></details>

I think we do need to reset it because we're directly modifying the server config, which means that later tests with the same server config would not spill to disk either. I would add a TODO to reenable this when we have recursive partitioning.
___
*[pkg/sql/colexec/execplan.go, line 586 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-KrSwd8WGjj1oJLt8z:-M-KrSwd8WGjj1oJLt9-:b-v98qco) ([raw file](https://github.com/cockroachdb/cockroach/blob/86a4930813d98573e5265784d853c7a13729ba60/pkg/sql/colexec/execplan.go#L586)):*
> ```Go
> 							// memory limit before creating the memory account for the
> 							// allocator because if we don't, the constructor below will
> 							// result an OOM panic.
> ```

nit: `s/an/in an`


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-:-M-L0dCr7RIK9vDbw4ke:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/execplan.go, line 576 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HQYcr02imHTtOdxI4:-M-L--ky34AtSAFvM26j:b-nlm76y) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L576)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Got it, might be useful to mention the recursive partitioning.
</blockquote></details>

Done. But even with recursive partitioning in-place we will need to do this reset.
___
*[pkg/sql/colexec/execplan.go, line 583 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-HR4shBSvG_4j5A3cz:-M-L-WyS65A3XMAHSalE:b-trm61h) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f74659197cd6a1dac390ba245777b81b4de1d5/pkg/sql/colexec/execplan.go#L583)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think we do need to reset it because we're directly modifying the server config, which means that later tests with the same server config would not spill to disk either. I would add a TODO to reenable this when we have recursive partitioning.
</blockquote></details>

I guess my comment came out wrong - I meant that not resetting the value to the original is wrong and should be fixed, but so far it didn't matter because in all of the tests we're setting memory limit right before calling `NewColOperator`. Well, maybe that's not true in case of `TestHashJoinerAgainstProcessor`. Anyway, fixed.

As I mentioned above, even with recursive partitioning we won't be able to remove because external hash joiner (at least currently) creates two `partitionerToOperator` operators each of which instantiates a batch of `coldata.BatchSize`. If we keep the memory limit unchanged, the constructor will panic because of an OOM panic from the allocator.
___
*[pkg/sql/colexec/execplan.go, line 586 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/44523#-M-KrSwd8WGjj1oJLt8z:-M-L0PyzAFuELaRv6okA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/86a4930813d98573e5265784d853c7a13729ba60/pkg/sql/colexec/execplan.go#L586)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: `s/an/in an`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r-

yuzefovich commented:
bors r+

asubiotto commented:
> I guess my comment came out wrong - I meant that not resetting the value to the original is wrong and should be fixed, but so far it didn't matter because in all of the tests we're setting memory limit right before calling NewColOperator. Well, maybe that's not true in case of TestHashJoinerAgainstProcessor. Anyway, fixed.

> As I mentioned above, even with recursive partitioning we won't be able to remove because external hash joiner (at least currently) creates two partitionerToOperator operators each of which instantiates a batch of coldata.BatchSize. If we keep the memory limit unchanged, the constructor will panic because of an OOM panic from the allocator.

I'm thinking more about logic tests. We're definitely going to have to figure something out here because logic tests assume that setting the `MemoryLimit` to 1 will force all processors to fall back to disk.
